Library/Homebrew/utils/github/api.rb

```diff
@@ -0,0 +1,304 @@
+        pretty_duration(Time.at(reset) - Time.now)
+      end
+    end
+
+    # Error when authentication fails.
+    class AuthenticationFailedError < Error
+      def initialize(github_message)
+        @github_message = github_message
+        message = +"GitHub #{github_message}:"
+        message << if Homebrew::EnvConfig.github_api_token
+          <<~EOS
+            HOMEBREW_GITHUB_API_TOKEN may be invalid or expired; check:
+              #{Formatter.url("https://github.com/settings/tokens")}
+          EOS
+        else
+          <<~EOS
+            The GitHub credentials in the macOS keychain may be invalid.
+            Clear them with:
+              printf "protocol=https\\nhost=github.com\\n" | git credential-osxkeychain erase
+            #{CREATE_GITHUB_PAT_MESSAGE}
+          EOS
+        end
+        super message.freeze
+      end
+    end
+
+    # Error when the user has no GitHub API credentials set at all (macOS keychain or envvar).
+    class MissingAuthenticationError < Error
+      def initialize
+        message = +"No GitHub credentials found in macOS Keychain or environment.\n"
+        message << CREATE_GITHUB_PAT_MESSAGE
+        super message
+      end
+    end
+
+    # Error when the API returns a validation error.
+    class ValidationFailedError < Error
+      def initialize(github_message, errors)
+        @github_message = if errors.empty?
+          github_message
+        else
+          "#{github_message}: #{errors}"
+        end
+
+        super(@github_message)
+      end
+    end
+
+    API_ERRORS = [
+      AuthenticationFailedError,
+      HTTPNotFoundError,
+      RateLimitExceededError,
+      Error,
+      JSON::ParserError,
+    ].freeze
+
+    # Gets the password field from `git-credential-osxkeychain` for github.com,
+    # but only if that password looks like a GitHub Personal Access Token.
+    sig { returns(T.nilable(String)) }
+    def keychain_username_password
+      github_credentials = Utils.popen(["git", "credential-osxkeychain", "get"], "w+") do |pipe|
+        pipe.write "protocol=https\nhost=github.com\n"
+        pipe.close_write
+        pipe.read
+      end
+      github_username = github_credentials[/username=(.+)/, 1]
+      github_password = github_credentials[/password=(.+)/, 1]
+      return unless github_username
+
+      # Don't use passwords from the keychain unless they look like
+      # GitHub Personal Access Tokens:
+      #   https://github.com/Homebrew/brew/issues/6862#issuecomment-572610344
+      return unless /^[a-f0-9]{40}$/i.match?(github_password)
+
+      github_password
+    rescue Errno::EPIPE
+      # The above invocation via `Utils.popen` can fail, causing the pipe to be
+      # prematurely closed (before we can write to it) and thus resulting in a
+      # broken pipe error. The root cause is usually a missing or malfunctioning
+      # `git-credential-osxkeychain` helper.
+      nil
+    end
+
+    def api_credentials
```

Comment:
Maybe others will disagree, but I think some of the `api_` prefixes can go if it's in an API module. e.g. `GitHub::API.api_credentials` seems unnecessary when it could be `GitHub::API.credentials`.

`open_api` could be `open_rest` to differentiate from `open_graphql`.